### PR TITLE
WIP: Include option to check SAML certificate expiration

### PR DESCRIPTION
**Pull request by CalRobert**
This allows a user to specify checkCertExpiration to force cert date validity checking. It will fail for expired certs as well as those not valid yet.

To be resolved:
* Do we need a pass on the x509 lib?
* Is server time something we can trust?
* We default-off on this (allowing expired certs), is that desired behaviour?

### CODE REVIEW

File: lib/passport-wsfed-saml2/saml.js [hunk 1]
```diff
@@ -119,7 +116,32 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
   }
 };
 
+SAML.prototype.validateCertExpiration = function (samlAssertion, version) {
```
Comment (robinbijlani) [inline lib/passport-wsfed-saml2/saml.js]: This logic checks the certificate embedded in the assertion, but I believe we also have a situation where if there is no embedded certificate, we will check a configured certificate (`options.cert`)

https://github.com/auth0/passport-wsfed-saml2/blob/7803d873fa64f3ff292a61f68085b54e9ae778ad/lib/passport-wsfed-saml2/saml.js#L58-L62

**Comment by CalRobert** [comment]
@robinbijlani  (is there a way to respond inline?)

* The cert period of validity is generally 3+ months, while the assertion's validity is ten minutes (at least looking through the tests), so I'm not sure there's a need to pad? Also if we are letting users enable this (see below) I'd want it to stick to a more strict interpretation

* In discussion on the zendesk ticket it sounds like this should be default-off for now. There's potential to cause a lot of pain. In that case, I don't think we'd need a change to API2 until such time as we decide we want default-on 

* Very true - have since refactored and am altering tests accordingly

**Comment by robinbijlani** [comment]
(You should be able to reply to each comment I left?)

I imagine the flag should be off for existing connections with expired certs (to not break existing auth flows), but should be turned on for new connections that are created. Otherwise, how would the checkCertExpiration flag ever be true?

Also, could you add a link to the Zendesk ticket in the description?

**Comment by CalRobert** [comment]
@robinbijlani Not sure re: options or feature flag. I think that's a product question (maybe this deserves a feature spec and is more than just a bugfix?) I'd also appreciate comment from @machuga 

Zendesk ticket is at https://auth0.zendesk.com/agent/tickets/45370

**Comment by CalRobert** [comment]
Incidentally, I have failing tests even on the master branch

  saml 1.1 assertion
...
    1) should fail when the X509Certificate is invalid

  saml 2.0 assertion
 ...
    2) should fail when the X509Certificate is invalid

**Comment by gkwang** [comment]
There should not be an option to turn off certificate validation checks. Is there a way to figure out what the impact for this if we enforce it strictly?

**Comment by CalRobert** [comment]
@gkwang as I understand it we've been ignoring certificate validity dates for 5 years already, though I agree it's bad. The concern is that breaking logins for people because their cert is expired is also bad. 

Personally, I think it makes sense to identify all the expired certs in use across our platform, find the expired ones, and go through CSM's to get companies to update. That is a slow, painful way to do things though. 

It sounds like this is getting to be outside the scope of sustainment. Perhaps a product manager should determine the best behaviour to implement?

File: lib/passport-wsfed-saml2/saml.js [hunk 2]
```diff
@@ -97,6 +96,12 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
     return callback(e);
   }
 
+  if (self.options.checkCertExpiration && !self.validateCertExpiration(xml)) {
+    valid = false;
+    return callback(new Error('SAML certificate not valid.'), null); //Note - could say "not currently valid" but
+                            // is it a risk to reveal that this is a date issue and not a generally invalid cert?
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: This is a valid point. Let's get another opinion on this.

File: lib/passport-wsfed-saml2/saml.js [hunk 3]
```diff
@@ -119,7 +124,34 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
   }
 };
 
+SAML.prototype.validateCertExpiration = function (samlAssertion) {
+  // TODO - is this version-dependent?
+  if(samlAssertion.getElementsByTagName("X509Certificate").length > 0){
+    cert = samlAssertion.getElementsByTagName("X509Certificate")[0].textContent;
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Is this var declared somewhere outside of scope or is it accidentally a global. You may want to assign it above the conditional and then reference the var name in the `if` so that you aren't traversing the XML twice.

File: lib/passport-wsfed-saml2/saml.js [hunk 4]
```diff
@@ -119,7 +124,34 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
...
+  } else {
+    cert = this.options.cert;
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Can you explain what this branch will do?

File: lib/passport-wsfed-saml2/saml.js [hunk 5]
```diff
@@ -119,7 +124,34 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
...
+  }
+
+  cert = utils.certToPEM(cert);
+  cert = x509.parseCert(cert)
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Rather than constantly reassigning this variable, it may make more sense to do some pipeline operations on it or just compose the functions together.

File: lib/passport-wsfed-saml2/saml.js [hunk 6]
```diff
@@ -119,7 +124,34 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
...
+
+  let nowDate = new Date();
+
+  if (cert.notBefore > nowDate) {
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: We can probably do a between check for these so we don't have two explicit checks + early returns.

File: lib/passport-wsfed-saml2/saml.js [hunk 7]
```diff
@@ -322,14 +354,11 @@ SAML.prototype.parseAssertion = function(samlAssertion, callback) {
...
     samlAssertion = samlAssertion.documentElement;
 
-  var version;
-  if (samlAssertion.getAttribute('MajorVersion') === '1')
-    version = '1.1';
-  else if (samlAssertion.getAttribute('Version') === '2.0')
-    version = '2.0';
-  else
+  var version = utils.getSamlAssertionVersion(samlAssertion);
+  let supported_versions = ['1.1', '2.0'];
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: This should be a `const` and it looks like the rest of the file mostly uses `camelCase` so I'd recommend this var match.

File: lib/passport-wsfed-saml2/saml.js (hunk 2 above)
Comment (nicosabena) [inline lib/passport-wsfed-saml2/saml.js]: If it is a risk to include that information in the response sent by the server, then we should at least put that information in a tenant log entry, as clearly as possible (preferably with start-end dates) so that customers can find what's wrong on their own.

File: lib/passport-wsfed-saml2/utils.js [hunk 8]
```diff
@@ -4,6 +4,20 @@ var SamlAssertionParserError = require('./errors/SamlAssertionParserError');
 var SamlResponseParserError = require('./errors/SamlResponseParserError');
 var WSFederationResultParserError = require('./errors/WSFederationResultParserError');
 
+exports.certToPEM = function (cert) {
+  cert = cert.match(/.{1,64}/g).join('\n');
```
Comment (machuga) [inline lib/passport-wsfed-saml2/utils.js]: I know the original version modifies the param passed in, but can we change the approach now that it is extracted?

Maybe even something like

```js
const CERT_START = "-----BEGIN CERTIFICATE-----\n";
const CERT_END = "\n-----END CERTIFICATE-----\n";
const certToPEM = (cert) => CERT_START + cert.match(/.{1,64}/g).join('\n') + CERT_END;
```
or 

```js
const CERT_START = "-----BEGIN CERTIFICATE-----";
const CERT_END = "-----END CERTIFICATE-----";
const certToPEM = (cert) => [].concat(
  CERT_START,
  cert.match(/.{1,64}/g).join('\n')
  CERT_END,
  '' // newline at end
).join("\n");
```

File: lib/passport-wsfed-saml2/utils.js [hunk 9]
```diff
@@ -4,6 +4,20 @@ var SamlAssertionParserError = require('./errors/SamlAssertionParserError');
...
+  cert = "-----BEGIN CERTIFICATE-----\n" + cert;
+  cert = cert + "\n-----END CERTIFICATE-----\n";
+  return cert;
+};
+
+exports.getSamlAssertionVersion = function(samlAssertion){
+  if (samlAssertion.getAttribute('MajorVersion') === '1')
+    return '1.1';
+  else if (samlAssertion.getAttribute('Version') === '2.0')
+    return '2.0';
```
Comment (machuga) [inline lib/passport-wsfed-saml2/utils.js]: Is `undefined` the desired behavior if the attrs do not match `1` or `2.0`?

File: lib/passport-wsfed-saml2/utils.js [hunk 10]
```diff
@@ -4,6 +4,20 @@ var SamlAssertionParserError = require('./errors/SamlAssertionParserError');
...
 
+exports.certToPEM = function (cert) {
+  cert = cert.match(/.{1,64}/g).join('\n');
+  cert = "-----BEGIN CERTIFICATE-----\n" + cert;
+  cert = cert + "\n-----END CERTIFICATE-----\n";
+  return cert;
+};
+
+exports.getSamlAssertionVersion = function(samlAssertion){
+  if (samlAssertion.getAttribute('MajorVersion') === '1')
```
Comment (machuga) [inline lib/passport-wsfed-saml2/utils.js]: Can you wrap these blocks in `{}`, please?

File: lib/passport-wsfed-saml2/saml.js [hunk 11]
```diff
@@ -21,19 +22,17 @@ var SAML = function (options) {
...
 };
 
-SAML.prototype.certToPEM = function (cert) {
-  cert = cert.match(/.{1,64}/g).join('\n');
-  cert = "-----BEGIN CERTIFICATE-----\n" + cert;
-  cert = cert + "\n-----END CERTIFICATE-----\n";
-  return cert;
-};
+// TODO - update all calls to this to simply use utils.
+SAML.prototype.certToPEM = utils.certToPEM;
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Rather than the TODO, let's just go ahead and replace these assuming the tests are in place.

File: test/saml20.tests.js [hunk 12]
```diff
@@ -23,6 +24,38 @@ describe('saml 2.0 assertion', function () {
     nameIdentifierFormat: 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified'
   };
 
+  it('should reject expired cert embedded in saml assertion', function(done){
+    // Turns out the cert we generate expired in 2013
+    let signedAssertion = saml20.create(options);
```
Comment (machuga) [inline test/saml20.tests.js]: Unless you plan on changing these later, favor `const`.

File: test/saml20.tests.js [hunk 13]
```diff
@@ -23,6 +24,38 @@ describe('saml 2.0 assertion', function () {
     nameIdentifierFormat: 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified'
   };
 
+  it('should reject expired cert embedded in saml assertion', function(done){
+    // Turns out the cert we generate expired in 2013
```
Comment (machuga) [inline test/saml20.tests.js]: Maybe rephrase or remove so maintainers don't start questioning what this is/was.

File: test/saml20.tests.js [hunk 14]
```diff
@@ -23,6 +24,38 @@ describe('saml 2.0 assertion', function () {
...
+    let signedAssertion = saml20.create(options);
+    let publicKey = fs.readFileSync(__dirname + '/test-auth0.cer').toString();
+    let saml_passport = new SamlPassport({cert: publicKey, realm: 'urn:myapp', checkRecipient: false, checkCertExpiration: true});
```
Comment (machuga) [inline test/saml20.tests.js]: Other variables are `camelCase`, I recommend switching these to `samlPassport`

File: test/saml20.tests.js [hunk 15]
```diff
@@ -23,6 +24,38 @@ describe('saml 2.0 assertion', function () {
...
+
+
+  it('should reject expired cert set in options', function(done){
+    // Turns out the cert we generate expired in 2013
+    let signedAssertion = saml20.create(options);
+    let publicKey = fs.readFileSync(__dirname + '/test-auth0.cer').toString();
+    let saml_passport = new SamlPassport({cert: publicKey, realm: 'urn:myapp', checkRecipient: false, checkCertExpiration: true});
+
+    // Kludgy, but now remove the cert
+    samlAssertion = utils.parseSamlResponse(signedAssertion);
```
Comment (machuga) [inline test/saml20.tests.js]: Global var - can you turn on linting in your editor to ensure all globals are caught?

File: test/saml20.tests.js [hunk 16]
```diff
@@ -23,6 +24,38 @@ describe('saml 2.0 assertion', function () {
     nameIdentifierFormat: 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified'
   };
 
+  it('should reject expired cert embedded in saml assertion', function(done){
+    // Turns out the cert we generate expired in 2013
+    let signedAssertion = saml20.create(options);
+    let publicKey = fs.readFileSync(__dirname + '/test-auth0.cer').toString();
```
Comment (machuga) [inline test/saml20.tests.js]: This could probably be moved to a `const` within this describe block since you reuse it.

File: lib/passport-wsfed-saml2/saml.js (hunk 2 above)
Comment (machuga) [reply lib/passport-wsfed-saml2/saml.js]: I think that's a solid direction. We'd need to ensure to attach some info to this error to bubble up to the service's error handlers.

Comment (gkwang) [reply lib/passport-wsfed-saml2/saml.js]: We should keep the external error message as is. I personally don't see any problems of mentioning the error being an expired certificate in the tenant log entry though. The cert start/end date may be unnecessary since the tenant should have access to the certificate and can check the expiration manually if need be.

File: lib/passport-wsfed-saml2/saml.js (hunk 11 above)
Comment (CalRobert) [reply lib/passport-wsfed-saml2/saml.js]: Done

**Comment by CalRobert** [comment]
@machuga responded to your comments. Can you resolve them if happy? Also was looking at the other errors reported and saying the cert isn't currently valid seems in line with the degree of detail included in other ones.

**Comment by CalRobert** [comment]
Note (particular to @robin)  the node4 tests fail due to use of const and let but it looks like that's true throughout `master` as well. Should we consider whether we still want to support Node4 with this?

**Comment by CalRobert** [comment]
Note (particular to @robinbijlani ) the node4 tests fail due to use of const and let but it looks like that's true throughout master as well. Should we consider whether we still want to support Node4 with this?

**Comment by CalRobert** [comment]
Note - waiting for confirmation that we don't need Node4 support.

**Comment by CalRobert** [comment]
Hi @machuga  - I think the only thing remaining is your review. I added changes in response to your comments; is there anything else you'd like?

File: lib/passport-wsfed-saml2/saml.js [hunk 17]
```diff
@@ -97,6 +93,11 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
     return callback(e);
   }
 
+  if (self.options.checkCertExpiration && !self.validateCertExpiration(xml)) {
+    valid = false;
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Does this serve any purpose if you return on the next line?

File: lib/passport-wsfed-saml2/utils.js [hunk 18]
```diff
@@ -4,6 +4,24 @@ var SamlAssertionParserError = require('./errors/SamlAssertionParserError');
 var SamlResponseParserError = require('./errors/SamlResponseParserError');
 var WSFederationResultParserError = require('./errors/WSFederationResultParserError');
 
+CERT_START = "-----BEGIN CERTIFICATE-----\n";
```
Comment (machuga) [inline lib/passport-wsfed-saml2/utils.js]: This has been declared global, please scope with `const`

File: lib/passport-wsfed-saml2/saml.js [hunk 19]
```diff
@@ -322,14 +337,11 @@ SAML.prototype.parseAssertion = function(samlAssertion, callback) {
...
 
-  var version;
-  if (samlAssertion.getAttribute('MajorVersion') === '1')
-    version = '1.1';
-  else if (samlAssertion.getAttribute('Version') === '2.0')
-    version = '2.0';
-  else
+  const version = utils.getSamlAssertionVersion(samlAssertion);
+  const supportedVersions = ['1.1', '2.0'];
+  if (supportedVersions.indexOf(version) === -1){
```
Comment (machuga) [inline lib/passport-wsfed-saml2/saml.js]: Since you have the method above checking if the version is known, can't you just check `if (version)` and remove the extra inclusion check?

File: lib/passport-wsfed-saml2/utils.js [hunk 20]
```diff
@@ -4,6 +4,24 @@ var SamlAssertionParserError = require('./errors/SamlAssertionParserError');
...
+const CERT_END = "\n-----END CERTIFICATE-----\n";
+const certToPEM = (cert) => CERT_START + cert.match(/.{1,64}/g).join('\n') + CERT_END;
+
+exports.certToPEM = certToPEM;
```
Comment (machuga) [inline lib/passport-wsfed-saml2/utils.js]: To match style you can probably just inline this and the method definition above.

File: lib/passport-wsfed-saml2/saml.js (hunk 17 above)
Comment (CalRobert) [reply lib/passport-wsfed-saml2/saml.js]: No, that's vestigial, thanks. removed.

File: lib/passport-wsfed-saml2/utils.js (hunk 18 above)
Comment (CalRobert) [reply lib/passport-wsfed-saml2/utils.js]: done

File: lib/passport-wsfed-saml2/saml.js (hunk 19 above)
Comment (CalRobert) [reply lib/passport-wsfed-saml2/saml.js]: The thinking was that we might update the util function with new versions, in whcih case `supportedVersions` doesn't carry any meaning. For instance, 6 months from now we return `2.2` from  `utils.getSamlAssertionVersion`


The old version had
```

  var version;
  if (samlAssertion.getAttribute('MajorVersion') === '1')
    version = '1.1';
  else if (samlAssertion.getAttribute('Version') === '2.0')
    version = '2.0';
  else
    return callback(new Error('SAML Assertion version not supported'), null);

```
and my thinking was to separate the logic that gets the version from the decision whether the version was supported. But I can change it to just check if the `version` is defined and remove the list of supported versions, of course.

File: lib/passport-wsfed-saml2/saml.js [hunk 21]
```diff
@@ -126,23 +123,22 @@ SAML.prototype.validateSignature = function (xml, options, callback) {
...
+    // This will use the cert embedded in the assertion if one is available.
+    cert = certNodes[0].textContent;
   } else {
+    // Otherwise, it uses the cert included in the options passed to the SAML constructor
     cert = this.options.cert;
   }
 
-  cert = utils.certToPEM(cert);
-  cert = x509.parseCert(cert)
+  cert = x509.parseCert(utils.certToPEM(cert));
```
[the remark this answers is not in the document]
Comment (CalRobert) [reply lib/passport-wsfed-saml2/saml.js]: Changed this to a separate `parsedCert` variable.

**Comment by CalRobert** [comment]
Responded to your comments @machuga - should be good now I think. Removed the separation of pulling version and deciding which versions are valid.

File: test/saml20.tests.js [hunk 22]
```diff
@@ -138,13 +139,75 @@ describe('saml 2.0 assertion', function () {
...
+
+    options.lifetimeInSeconds = 10000;
+
+    var signedAssertion = saml20.create(options);
+    var publicKey = fs.readFileSync(__dirname + '/test-auth0.cer').toString();
+    // The embedded cert is expired, so we can use this as is.
+    const samlPassport = new SamlPassport({cert: publicKey, realm: 'urn:myapp', checkRecipient: false});
+    samlPassport.validateSamlAssertion(signedAssertion, function(err, profile) {
+      should.not.exists(err);
+      // err.message.should.equal('The signing certificate is not currently valid.');
```
Comment (machuga) [inline test/saml20.tests.js]: Any reason for this to hang out?

Comment (CalRobert) [reply test/saml20.tests.js]: No, just some copy-pasted cruft. Removing now. Thanks for catching that.

Comment (CalRobert) [reply test/saml20.tests.js]: Removed (and resolving, if that's not presumptuous on my part)
